### PR TITLE
Tweak the API of the `Val` type

### PR DESCRIPTION
**Pull request by alexcrichton**
A few updates to the API of the `Val` type:

* Added a payload for `V128`.
* Replace existing accessor methods with `Option`-returning versions.
* Add `unwrap_xxx` family of methods to extract a value and panic.
* Remove `Into` conversions which panic, since panicking in `From` or
  `Into` isn't idiomatic in Rust
* Add documentation to all methods/values/enums/etc.
* Rename `Val::default` to `Val::null`

### CODE REVIEW

File: crates/interface-types/src/lib.rs [hunk 1]
```diff
@@ -152,7 +152,13 @@ impl ModuleData {
             Ok(values) => values
                 .to_vec()
                 .into_iter()
-                .map(|v: wasmtime::Val| v.into())
+                .map(|v: wasmtime::Val| match v {
+                    wasmtime::Val::I32(i) => RuntimeValue::I32(i),
+                    wasmtime::Val::I64(i) => RuntimeValue::I64(i),
+                    wasmtime::Val::F32(i) => RuntimeValue::F32(i),
+                    wasmtime::Val::F64(i) => RuntimeValue::F64(i),
```
Comment (abrown) [inline crates/interface-types/src/lib.rs]: No `V128`?

File: crates/api/src/values.rs [hunk 2]
```diff
@@ -52,52 +103,35 @@ impl Val {
...
 
-    pub fn from_f32_bits(v: u32) -> Val {
-        Val::F32(v)
+    accessors! {
+        e
+        (I32(i32) i32 unwrap_i32 *e)
+        (I64(i64) i64 unwrap_i64 *e)
+        (F32(f32) f32 unwrap_f32 f32::from_bits(*e))
+        (F64(f64) f64 unwrap_f64 f64::from_bits(*e))
+        (FuncRef(&HostRef<Func>) funcref unwrap_funcref e)
```
Comment (abrown) [inline crates/api/src/values.rs]: Do we need a `V128` line here as well?

File: crates/api/src/values.rs [hunk 3]
```diff
@@ -6,21 +6,71 @@ use std::ptr;
 use wasmtime_environ::ir;
 use wasmtime_jit::RuntimeValue;
 
+/// Possible runtime values that a WebAssembly module can either consume or
+/// produce.
```
Comment (abrown) [inline crates/api/src/values.rs]: Perhaps we should describe here and in `RuntimeValue` what is the difference between these `Val` and `RuntimeValue`. I'm not very clear on their differences and judging superficially from some of the overlap (e.g. `RuntimeValue::unwrap_*`) I wonder if we shouldn't just create one version of a runtime value struct and move it to a place where both crates can share the definition.

File: crates/api/src/values.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/api/src/values.rs]: Oops, indeed!

File: crates/interface-types/src/lib.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/interface-types/src/lib.rs]: Oops, forgot!

File: crates/api/src/values.rs (hunk 3 above)
Comment (alexcrichton) [reply crates/api/src/values.rs]: Hm I'm not sure if this would be the best place for that, but it sort of depends on the user. The `wasmtime` crate is still relatively new and idioms are still changing, but I believe the intention is that the `wasmtime` crate is almost exclusively used when embedding wasmtime somewhere else, unless a much more advanced use case is required.

The `RuntimeValue` type, AFAIK, is purely wasmtime-internals in that it's part of the nitty-gritty details of `wasmtime`'s own internal implementation. The `Val` type here, though, is intended to be user-facing and what you actually interact with. I don't think users are expected to interact with `RuntimeValue` at all if they're using the `wasmtime` crate. 

So basically I think that this won't be a problem as time goes on because users won't ever even see `RuntimeValue`, and otherwise proposals like https://github.com/bytecodealliance/wasmtime/issues/677 may make unifying impossible since the `wasmtime::Val` will be richer than `RuntimeValue`.
